nemo/collections/asr/parts/perturb.py

```diff
@@ -20,20 +27,164 @@ def perturb(self, data):
 
 
 class SpeedPerturbation(Perturbation):
-    def __init__(self, min_speed_rate=0.85, max_speed_rate=1.15, rng=None):
+    def __init__(self, sr, min_speed_rate=0.9, max_speed_rate=1.1, num_rates=5, rng=None):
+        """
+        Performs Speed Augmentation by re-sampling the data to a different sampling rate,
+        which does not preserve pitch.
+
+        Note: This is a very slow operation for online augmentation. If space allows,
+        it is preferable to pre-compute and save the files to augment the dataset.
+
+        Args:
+            sr: Original sampling rate.
+            min_speed_rate: Minimum sampling rate modifier.
+            max_speed_rate: Maximum sampling rate modifier.
+            num_rates: Number of discrete rates to allow. Can be a positive or negative
+                integer.
+                If a positive integer greater than 0 is provided, the range of
+                speed rates will be discretized into `num_rates` values.
+                If a negative integer or 0 is provided, the full range of speed rates
+                will be sampled uniformly.
+                Note: If a positive integer is provided and the resultant discretized
+                range of rates contains the value '1.0', then those samples with rate=1.0,
+                will not be augmented at all and simply skipped. This is to unnecessary
+                augmentation and increase computation time. Effective augmentation chance
+                in such a case is = `prob * (num_rates - 1 / num_rates) * 100`% chance
+                where `prob` is the global probability of a sample being augmented.
+            rng: Random seed number.
+        """
+        self._sr = sr
         self._min_rate = min_speed_rate
         self._max_rate = max_speed_rate
+        self._num_rates = num_rates
+        if num_rates > 0:
+            self._rates = np.linspace(self._min_rate, self._max_rate, self._num_rates, endpoint=True)
         self._rng = random.Random() if rng is None else rng
 
     def max_augmentation_length(self, length):
         return length * self._max_rate
 
     def perturb(self, data):
-        speed_rate = self._rng.uniform(self._min_rate, self._max_rate)
+        # Select speed rate either from choice or random sample
+        if self._num_rates < 0:
+            speed_rate = self._rng.uniform(self._min_rate, self._max_rate)
+        else:
+            speed_rate = self._rng.choice(self._rates)
+
+        # Skip perturbation in case of identity speed rate
+        if speed_rate == 1.0:
+            return
+
         if speed_rate <= 0:
             raise ValueError("speed_rate should be greater than zero.")
-        # logging.debug("speed: %f", speed_rate)
-        data._samples = librosa.effects.time_stretch(data._samples, speed_rate)
+
+        new_sr = int(self._sr * speed_rate)
+        data._samples = librosa.core.resample(data._samples, self._sr, new_sr, res_type='kaiser_fast')
```

Comment:
I would hesitate to default to kasier_fast, this would result in degraded ASR training.

Reply:
I was thinking of speed over quality. I'll expose it as a parameter, and default to Kaiser fast. Is that fine?

Kaiser best runtime cost is 2.7x on average that of Kaiser fast, which already takes over 100 ms per file of maximum duration 16.7 seconds in my speed tests.

Reply:
This is part of the reason we tend to shy away from doing online speed perturbation.
I'm not too sure about defaulting to kaiser_fast.

Reply:
So with kaiser fast, I could increase the range of augmentations from [90%, 100%, 110%] to [90%, 95%, 100%, 105%, 110%] with a 2x slowdown of training. If we augment the data offline, the training time is technically 3x normal time. So there is some benefit to on the fly augmentation, even if its marginal. 

As to kaiser_fast vs kaiser_best, its a toss up. Personally id go for speed over quality, but making it a user choice is probably the best. Depending on amount of data and quality of original data, either option might be better.

Reply:
How about for ASR?

Let's just make a required argument.

Reply:
Sounds good.

Reply:
Made it a required arg